query/src/main/scala/filodb/query/exec/rangefn/RateFunctions.scala

```diff
@@ -301,3 +399,154 @@ class HistIncreaseFunction extends HistogramRateFunctionBase {
+        highestTime = endTime
+      }
+      sum += dblReader.sum(acc, vector, start = startRowNum, end = endRowNum) // Sum of the values gives us the increase
+    }
+  }
+
+  override def apply(windowStart: Long, windowEnd: Long, sampleToEmit: TransientRow): Unit = {
+    if (highestTime > lowestTime) {
+      // NOTE: It seems in order to match previous code, we have to adjust the windowStart by -1 so it's "inclusive"
+      val result = RateFunctions.extrapolatedPeriodicRate(
+        windowStart - 1, windowEnd, numSamples,
+        lowestTime,
+        highestTime,
+        sum,
+        isRate)
+      sampleToEmit.setValues(windowEnd, result)
+    } else {
+      sampleToEmit.setValues(windowEnd, Double.NaN)
+    }
+  }
+
+  def apply(endTimestamp: Long, sampleToEmit: TransientRow): Unit = ???
+}
+
+class ChunkedPeriodicRateFunction extends ChunkedPeriodicRateFunctionBase {
+  def isRate: Boolean    = true
+}
+
+class ChunkedPeriodicIncreaseFunction extends ChunkedPeriodicRateFunctionBase {
+  def isRate: Boolean    = false
+}
+
+
+/**
+ * A base class for chunked calculation of rate etc for period-counter-like histograms.
+ * Note that the rate of two histograms is itself a histogram.
+ * Similar algorithm to ChunkedPeriodicRateFunctionBase.
+ * It is O(nWindows * nChunks) which is usually << O(nSamples).
+ */
+abstract class HistogramPeriodicRateFunctionBase extends PeriodCounterChunkedRangeFunction[TransientHistRow] {
+  var numSamples = 0
+  var lowestTime = Long.MaxValue
+  var highestTime = 0L
+  var summedHist: bv.MutableHistogram = bv.MutableHistogram.empty(HistogramBuckets.emptyBuckets)
+
+  def isRate: Boolean
+
+  override def reset(): Unit = {
+    numSamples = 0
+    lowestTime = Long.MaxValue
+    highestTime = 0L
+    summedHist = bv.MutableHistogram.empty(HistogramBuckets.emptyBuckets)
+    super.reset()
+  }
+
+  def addTimeChunks(acc: MemoryReader, vector: BinaryVectorPtr, reader: VectorDataReader,
+                    startRowNum: Int, endRowNum: Int,
+                      startTime: Long, endTime: Long): Unit = reader match {
+    case histReader: bv.RowHistogramReader =>
+      if (startTime < lowestTime || endTime > highestTime) {
+        numSamples += endRowNum - startRowNum + 1
+        if (startTime < lowestTime) {
+          lowestTime = startTime
+        }
+        if (endTime > highestTime) {
+          highestTime = endTime
+        }
+        if (summedHist.isEmpty) {
+          summedHist = histReader.sum(startRowNum, endRowNum)
+        } else {
+          summedHist.addNoCorrection(histReader.sum(startRowNum, endRowNum))
+        }
+      }
+    case _ =>
+  }
+
+  override def apply(windowStart: Long, windowEnd: Long, sampleToEmit: TransientHistRow): Unit = {
+    if (highestTime > lowestTime) {
+      // NOTE: It seems in order to match previous code, we have to adjust the windowStart by -1 so it's "inclusive"
+      // TODO: handle case where schemas are different and we need to interpolate schemas
```

Comment:
Should we catch this case now and throw an exception?

Reply:
modified the implementation